src/python/pants/backend/helm/dependency_inference/deployment.py

```diff
@@ -0,0 +1,180 @@
+            field_set=field_set,
+            description=f"Rendering Helm deployment {field_set.address}",
+        ),
+    )
+
+    parsed_manifests = await MultiGet(
+        Get(
+            ParsedKubeManifest,
+            ParseKubeManifestRequest(filename=file, digest=rendered_deployment.snapshot.digest),
+        )
+        for file in rendered_deployment.snapshot.files
+    )
+
+    # Build YAML index of `ImageRef`s for future processing during depedendecy inference or post-rendering.
+    image_refs_index: MutableYamlIndex[ImageRef] = MutableYamlIndex()
+    for manifest in parsed_manifests:
+        for (idx, path, image_ref) in manifest.found_image_refs:
+            image_refs_index.insert(
+                file_path=PurePath(manifest.filename),
+                document_index=idx,
+                yaml_path=path,
+                item=ImageRef.parse(image_ref),
+            )
+
+    return HelmDeploymentReport(address=field_set.address, image_refs=image_refs_index.frozen())
+
+
+@dataclass(frozen=True)
+class FirstPartyHelmDeploymentMappings:
+    deployment_to_docker_addresses: FrozenDict[Address, FrozenYamlIndex[tuple[ImageRef, Address]]]
+
+    def referenced_by(self, address: Address) -> list[tuple[ImageRef, Address]]:
+        if address not in self.deployment_to_docker_addresses:
+            return []
+        return list(self.deployment_to_docker_addresses[address].values())
+
+
+@rule
+async def first_party_helm_deployment_mappings(
+    deployment_targets: AllHelmDeploymentTargets, docker_targets: AllDockerImageTargets
+) -> FirstPartyHelmDeploymentMappings:
+    field_sets = [HelmDeploymentFieldSet.create(tgt) for tgt in deployment_targets]
+    all_deployments_info = await MultiGet(
+        Get(HelmDeploymentReport, HelmDeploymentFieldSet, field_set) for field_set in field_sets
+    )
+
+    docker_target_addresses = {tgt.address.spec: tgt.address for tgt in docker_targets}
+
+    def lookup_docker_addreses(image_ref: ImageRef) -> tuple[ImageRef, Address] | None:
+        addr = docker_target_addresses.get(str(image_ref), None)
+        if addr:
+            return image_ref, addr
+        return None
+
+    # Builds a mapping between `helm_deployment` addresses and a YAML index of `docker_image` addresses.
+    address_mapping = {
+        fs.address: info.image_refs.transform_values(lookup_docker_addreses)
+        for fs, info in zip(field_sets, all_deployments_info)
+    }
+    return FirstPartyHelmDeploymentMappings(
+        deployment_to_docker_addresses=FrozenDict(address_mapping)
+    )
+
+
+@dataclass(frozen=True)
+class HelmDeploymentDependenciesInferenceFieldSet(FieldSet):
+    required_fields = (HelmDeploymentDependenciesField,)
+
+    dependencies: HelmDeploymentDependenciesField
+
+
+class InferHelmDeploymentDependenciesRequest(InferDependenciesRequest):
+    infer_from = HelmDeploymentDependenciesInferenceFieldSet
+
+
+@rule(desc="Find the dependencies needed by a Helm deployment", level=LogLevel.DEBUG)
```

Comment:
The underlying process doing the extraction is likely to have a better description (since it will include the filename), so I'd remove the `level` here and let this default to TRACE.